### PR TITLE
Search bar on dashboard

### PR DESCRIPTION
**Pull request by b3njamint**
### Summary 

This pull request implements a search bar on the dashboard to search for courses. Looks for course names that contain your search string as a substring, including cross-listed courses.

Edit: Fixed merge problems with @jewang25's pull request. Also unified designs of the various search features.

- [x] implemented X
- [ ] fixed Y

### Test Plan 

When you type in "C", shows only courses with "C" in the name.
Screen Shot 2022-10-06 at 3 41 55 PM

When you type in "CS 1", shows only courses with "CS 1" in the name.
Screen Shot 2022-10-06 at 3 42 12 PM

**Test for cross-listed courses.** When you type in "ENGR", shows only courses with "ENGR" in the name. Note that CS 2110 is cross-listed as ENGRD 2110 (so is ECE 2300 and ENGRD 2300 but no students submitted a form for ENGRD 2300 yet).
Screen Shot 2022-10-06 at 3 42 23 PM

### Breaking Changes 

Don't think there are any, as this is all done in the frontend (does not touch Firebase).

### CODE REVIEW

**Comment by jewang25** [comment]
The search bar looks good! I like how when I tried adding multiple courses that start with the same letter, all of them show up and I can add letters to the search to narrow down the courses that show up. I also like how you combined searching with the sort option selected, but we'll also have to add it for filtering once that pr gets merged in. It's also nice how you were able to integrate the new code for the search bar with the existing code without making too many changes to the code.

**Review by jewang25 — COMMENTED** [review]
The search bar looks good! I like how when I tried adding multiple courses that start with the same letter, all of them show up and I can add letters to the search to narrow down the courses that show up. I also like how you combined searching with the sort option selected, but we'll also have to add it for filtering once that pr gets merged in. It's also nice how you were able to integrate the new code for the search bar with the existing code without making too many changes to the code.

**Review by Coreball — COMMENTED** [review]
I think when you attempted to fix the merge conflicts most recently, you deleted the code @jewang25 added in #103 instead of preserving it. You should fix this because otherwise this PR will erase those changes when merged.

**Comment by b3njamint** [comment]
> I think when you attempted to fix the merge conflicts most recently, you deleted the code @jewang25 added in #103 instead of preserving it. You should fix this because otherwise this PR will erase those changes when merged.

File: frontend/src/modules/Dashboard/Components/Dashboard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/Dashboard.tsx]: I think the variable name for the search query could be more descriptive than `message`. You can also remove the console log here because it probably won't be useful other than debugging once we know it works

File: frontend/src/modules/Dashboard/Components/Dashboard.tsx
Comment (Coreball) [inline frontend/src/modules/Dashboard/Components/Dashboard.tsx]: Small thing, but I think it would be clearer to use `.some()` instead of `.find()` here for filtering because `.some()` explicitly returns a boolean whether it satisfies the predicate or not while `.find()` returns the value. You can also omit the curly braces and return statement and use the expression directly.

**Review by Coreball — APPROVED** [review]
Nice work! Thanks for implementing the search feature, it's very helpful because there can be hundreds of courses to scroll through while trying to find a specific one. I am approving this because it works, but there are just a few minor nits in my review that you should take a look at. Good job fixing the merge conflicts with @jewang25 's PR and adapting the code here to work with that in a unified visual style.

**Comment by b3njamint** [comment]
Before I merge, I want to check in if I deleted anything important when fixing the merge conflicts (again). I think the conflicting code was from an old version that someone probably had locally and pushed.

**Comment by mluo24** [comment]
@b3njamint You unfortunately have some errors going on, which are probably a result of the merge. It's a good idea to try and run your code before pushing so that you at least can see if there are errors and if it will run. 

I also think you also deleted some of the code. Jessica made a new PR with changes to the same files so I think that's why it looks a little different. The link to your [files](https://github.com/cornell-dti/zing-lsc/pull/107/files) and also the link to Jessica's [PR 1](https://github.com/cornell-dti/zing-lsc/pull/103) and [PR 2](https://github.com/cornell-dti/zing-lsc/pull/113) should help you in deciding what code you deleted so you can put it back. 

Once you fix this, lmk and you should be good to go!

**Comment by b3njamint** [comment]
I think #113 merged between when I fixed the merge conflicts last time and now, so it caused more conflicts. They should be fixed now @mluo24.

**Comment by mluo24** [comment]
@b3njamint unfortunately I think you still missed a few things, even from #113, and you're missing a few of Justin's changes from his PRs. I recommend you to try again with the merging, but before you do that, read your PR diff in the "files changed" tab currently and see if it makes sense to have deleted what you deleted. If you don't remember deleting the line or you didn't have a reason to, it's probably something you should keep since it's a change from main that you need to incorporate.

**Comment by b3njamint** [comment]
Thanks for the feedback @mluo24! I read through the files changed and put the deleted lines back accordingly.

**Comment by JustKong13** [comment]
Everything with the search bar looks good so far! I like how you are able to search course numbers as well now. Just one thing, you wrote that classes that are cross-listed also pop up when you search up their non-default course name. It doesn't seem to be working for me when I type in "ENGRD", even after I submitted surveys with that cross-listed course.

**Comment by JustKong13** [comment]
> Everything with the search bar looks good so far! I like how you are able to search course numbers as well now. Just one thing, you wrote that classes that are cross-listed also pop up when you search up their non-default course name. It doesn't seem to be working for me when I type in "ENGRD", even after I submitted surveys with that cross-listed course.

Nvm, I think I misspelled ENGRD or something, it's working now. Good job!

**Comment by mluo24** [comment]
There's not exactly a 1:1 mapping (a few spaces added here and there, changed variables names), but it's good enough since the functionality should be all preserved. Feel free to merge it into main! Good job again :) @b3njamint

**Comment by hexbabe** [comment]
yasss

**Pull request merged**
